tests/rptest/services/redpanda.py

```diff
@@ -1484,21 +1473,63 @@ def schema_reg(self, limit=None) -> str:
         ]
         return ",".join(schema_reg)
 
-    def metrics(self, node):
+    def metrics(self, node) -> pandas.DataFrame:
+        """
+        Returns the current metrics snapshot as a pandas DataFrame. Each metric sample
+        is flattened into a single DataFrame row with the following format.
+
+        (family, name, value, timestamp, label1, label2.....).
+
+        Label columns are a union of labels across all the metric samples. Since labels
+        are not uniformly typed across samples, pandas treats them as strings and should be
+        queried accordingly.
+
+        Example:
+
+        vectorized_alien_receive_batch_queue_length{shard="0"} 0.000000
+        vectorized_httpd_connections_total{service="admin",shard="0"} 3
+
+        translates to
+
+                                                family                                         name  value timestamp shard service
+        0  vectorized_alien_receive_batch_queue_length  vectorized_alien_receive_batch_queue_length    0.0      None     0    None
+        1           vectorized_httpd_connections_total           vectorized_httpd_connections_total   31.0      None     0   admin
+
+
+        Callers can use query abstractions like pandasql for complex filtering and aggregations.
+
+        Example: Find total bytes read by reactor threads across all shards on a node.
+
+        query_metrics(node, "select sum(value) from df where name='vectorized_reactor_aio_bytes_read_total'")
+        """
         assert node in self._started
         url = f"http://{node.account.hostname}:9644/metrics"
         resp = requests.get(url)
         assert resp.status_code == 200
-        return text_string_to_metric_families(resp.text)
 
-    def metrics_sample(self,
-                       sample_pattern,
-                       nodes=None) -> Optional[MetricSamples]:
+        metric_families = text_string_to_metric_families(resp.text)
```

Comment:
This seems a bit higher-overhead than the old code: previously we were doing the transformation once in text_string_to_metrics_families, then discarding anything that wasn't in `sample_pattern`.  Now, the whole body of all metrics is being carried through + transformed into a data frame.

Why not cull the unwanted metrics earlier?

Reply:
`query_metrics(sql)` below does that? It transforms this "all samples" data frame into a leaner one based on the filtering logic in the query. or did I misunderstand your question?

Reply:
Right, we eventually cull the unwanted ones, but it's quite late in the process, after we've already transformed everything a couple of times (from text to dict, to another dict, to a dataframe).  It seems like we could be quite a bit more efficient by culling the unwanted ones earlier, e.g. before inserting them into metrics_samples.

This stuff will not be noticeably slow until we try to use it on tests that have larger partition counts, at which point the overhead of a python loop+insert across 100k metrics could become problematic.

Reply:
(missed this comment)

ya that makes sense, one of my TODOs in this PR is check the performance implications on the test runtime by measuring at scale. If you think the direction in the PR is good and worth the effort, I can continue on to do performance tests. If not I can close this one out.